### DISCUSSION TITLE
v3.0.0-rc.5

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17183] [Announcements]
[Check the Release Notes](https://github.com/nuxt/framework/releases/tag/v3.0.0-rc.5)

**Comment by LoiLock** [comment]
Ran into https://github.com/nuxt/framework/pull/5874 today, perfect!

**Comment by leonardssh** [comment]
![MotiHeartsGIF](https://user-images.githubusercontent.com/35312043/178791514-8290368d-8346-42c4-bba4-8b852d4d82ea.gif)

**Comment by jshimkoski** [comment]
Did something change related to how `useRequestHeaders()` works? The upgrade to RC 5 made my app go kablooie.

Long story short, I have a custom auth plugin that uses `useRequestHeaders()` to grab the headers (think basic auth, other SSO login options). What used to be populated with headers is now just this value: `{ host: undefined }`.

**Reply by pi0** [reply]
Nothing in particular but might be a regression of presets. Does it also happen with `nuxt dev` to be broken? How is your usage?

**Reply by jshimkoski** [reply]
Yes. It breaks when using both `npm run dev` and `npm run build`. This is part of the error log that I'm seeing:

```bash
[Vue warn]: Component <Anonymous> is missing template or render function.
[nuxt] [request error] Cannot read properties of null (reading 'value')
  at ReactiveEffect.fn (./.output/server/chunks/app/server.mjs:55842:23)  
  at ReactiveEffect.run (./.output/server/chunks/handlers/renderer.mjs:1389:25)  
  at ComputedRefImpl.get value [as value] (./.output/server/chunks/handlers/renderer.mjs:2277:39)  
  at Object.unref (./.output/server/chunks/handlers/renderer.mjs:2188:29)  
  at ./.output/server/chunks/app/server.mjs:55869:91  
  at renderComponentSubTree (./.output/server/chunks/handlers/renderer.mjs:10591:13)  
  at ./.output/server/chunks/handlers/renderer.mjs:10523:29  
  at processTicksAndRejections (node:internal/process/task_queues:96:5)  
  at async unrollBuffer (./.output/server/chunks/handlers/renderer.mjs:10743:24)  
  at async unrollBuffer (./.output/server/chunks/handlers/renderer.mjs:10749:24)
[nitro] [dev] [unhandledRejection] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[Vue warn]: Component <Anonymous> is missing template or render function.
[nuxt] [request error] Cannot read properties of null (reading 'value')
  at ReactiveEffect.fn (./.output/server/chunks/app/server.mjs:55842:23)  
  at ReactiveEffect.run (./.output/server/chunks/handlers/renderer.mjs:1389:25)  
  at ComputedRefImpl.get value [as value] (./.output/server/chunks/handlers/renderer.mjs:2277:39)  
  at Object.unref (./.output/server/chunks/handlers/renderer.mjs:2188:29)  
  at ./.output/server/chunks/app/server.mjs:55869:91  
  at renderComponentSubTree (./.output/server/chunks/handlers/renderer.mjs:10591:13)  
  at ./.output/server/chunks/handlers/renderer.mjs:10523:29  
  at processTicksAndRejections (node:internal/process/task_queues:96:5)  
  at async unrollBuffer (./.output/server/chunks/handlers/renderer.mjs:10743:24)  
  at async unrollBuffer (./.output/server/chunks/handlers/renderer.mjs:10749:24)
[nitro] Error while generating error response FetchError: 302  (/__nuxt_error?url=/login&statusCode=500&statusMessage=Internal+Server+Error&message=Cannot+read+properties+of+null+(reading+'value')&description&data)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.errorhandler [as onError] (<REDACTED>/.output/server/chunks/nitro/node-server.mjs:322:16)
    at async Server.nodeHandler (<REDACTED>/.output/server/node_modules/h3/dist/index.mjs:373:9)
```

Here is a subset of the code for my auth plugin:

```javascript
export default defineNuxtPlugin(() => {
  const token = useCookie('token')
  const isLoggedIn = useState('isLoggedIn', () => !!token.value)
  const user = useState<Partial<IAuthUser>>('user', () => null)
  
  const { API_URL: baseURL } = useRuntimeConfig()

  addRouteMiddleware('global-auth', async (to) => {
    if (!to.path.includes('/login')) {
      try {
        await login()
      } catch (e) {
        return logout()
      }
    } else if (isLoggedIn.value && to.path.includes('/login')) {
      try {
        await login()
        return '/'
      } catch (e) {
        return logout()
      }
    }
  }, { global: true })

  const getTokenViaSSO = async () => {
    try {
      const reqHeaders = useRequestHeaders()
      const headers = []
      const method = { name: null }

      if (reqHeaders.authorization) {
        method.name = 'basic'
        headers.push(...['authorization'])
      }

      if (!token.value) {
        const data = await $fetch<any>('/users/login', {
          baseURL,
          method: 'post',
          onRequest({ options }) {
            if (process.server) {
              options.headers = useRequestHeaders(headers);
            }
            return Promise.resolve()
          },
          body: { method }
        })
        token.value = data.token
      }
    } catch(e) {
      throw Error(e)
    }
  }

  const getTokenViaForm = async (username: string, password: string) => {
    try {
      if (!token.value) {
        const data = await $fetch<any>('/users/login', {
          baseURL,
          method: 'post',
          body: {
            method: { name: 'form' },
            username,
            password
          }
        })
        token.value = data.token
      }
    } catch(e) {
      throw Error(e)
    }
  }

  const getUser = async () => {
    try {
      const data = await $fetch<any>('/users/me', {
        baseURL,
        method: 'get',
        headers: {
          'x-token': token.value
        },
      })
      user.value = data
      isLoggedIn.value = true
    } catch(e) {
      throw Error(e)
    }
  }

  const resetAuthValues = () => {
    token.value = null
    user.value = null
    isLoggedIn.value = false
  }

  const logout = async () => {
    resetAuthValues()
    return navigateTo('/login')
  }

  const login = async (username?: string, password?: string) => {
    try {
      if (username && password) {
        if (!token.value) await getTokenViaForm(username, password)
      } else {
        if (!token.value) await getTokenViaSSO()
      }
      if (!user.value) await getUser()
      if (username && password) return navigateTo('/')
    } catch (e) {
      throw Error(e)
    }
  }

  return {
    provide: {
      auth: {
        token,
        isLoggedIn,
        user,
        getTokenViaForm,
        getTokenViaSSO,
        getUser,
        login,
        logout
      }
    }
  }
})
```

Again, this all worked great in RC 4 and below. RC 5 comes, and kaboom.

**Reply by pi0** [reply]
Seems related to https://github.com/nuxt/framework/pull/5573. Can you please check?

**Reply by jshimkoski** [reply]
Am I reading correctly that #5573 will only break if there is a `.value` added to a ref inside of a template?

Just want to double check that I read that correctly.

**Reply by jshimkoski** [reply]
Moved this discussion to a new issue: https://github.com/nuxt/framework/issues/5890

**Comment by Amirali-Yavari** [comment]
Thanks
please update the documents

**Reply by pi0** [reply]
Which part of the docs do you think can be improved?

**Reply by Amirali-Yavari** [reply]
**Module Author Guide** 
i think this part is not can be improved and nuxt modules structure can explain better

**Reply by pi0** [reply]
Sure. Thanks for sharing feedback. Can you explain what items are you thinking about to explain better module structure? (BTW PR is more than welcome to improve. Would be happy to help explaining any missing information)

**Reply by Amirali-Yavari** [reply]
when you install with 
`npx nuxi init -t module my-module`
have folder named  `playground`
and in docs we havent got any explain about this 
or `src` folder we dont know what file have to be in this folder and finally please add summary for all folder and files in this structure

**Comment by shtse8** [comment]
after upgrading, my app crashed .

```
Identifier '$id_1681639291' has already been declared
```

After spending hours to investigate, the issue is caused by vuetify plugin. But it is working fine in rc4.

```ts
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'

export default defineNuxtPlugin((nuxtApp) => {
  const plugin = createVuetify({
    components,
    directives,
    theme: false,
  })

  nuxtApp.vueApp.use(plugin)
})
```

**Reply by hanscahya** [reply]
`[nuxt] [request error] Identifier '$id_2982235429' has already been declared
  at Loader.moduleStrategy (internal/modules/esm/translators.js:145:18)  
  at async link (internal/modules/esm/module_job.js:67:21)`

Got the same error here, not sure why

**Reply by mamlzy** [reply]
[nuxt] [request error] Identifier '$id_944007393' has already been declared

after upgrading, i got the same error too, how to fix this? @pi0

**Reply by shtse8** [reply]
Don't use vuetify, then everything will be solved.

**Reply by pi0** [reply]
It is likely a hash conflict. Do you mind to create a reproduction i can try? 🙏

**Reply by mamlzy** [reply]
> Don't use vuetify, then everything will be solved.

i am not even use vuetify, but still get the error

**Reply by shtse8** [reply]
> It is likely a hash conflict. Do you mind to create a reproduction i can try? 🙏

Sure. I will create a reproduction to you.

**Reply by shtse8** [reply]
> > Don't use vuetify, then everything will be solved.
> 
> i am not even use vuetify, but still get the error

Then probably, some others modules are broken.

**Reply by pi0** [reply]
Related issue: https://github.com/nuxt/framework/issues/5889

**Reply by misaon** [reply]
reproduction with vue-toastification here: https://stackblitz.com/edit/github-gysihc?file=nuxt.config.ts (https://github.com/nuxt/framework/issues/5889)

**Comment by warmthsea** [comment]
When is the official release expected?
预计什么时候发布正式版？

**Comment by QaahirStewart** [comment]
getting on RC5 not 4 : [nuxt] You are using a plugin that has not been wrapped in `defineNuxtPlugin`. It is advised to wrap your plugins as in the future this may enable enhancements: how to fix this

**Reply by danielroe** [reply]
It's either a plugin you're adding (in which case you can fix it by adding `defineNuxtPlugin`) or a module you're using.

**Reply by QaahirStewart** [reply]
ok found out my unocss module wasn't @latest, updated running w/o the error now. thanks

**Reply by lastmirage** [reply]
can nuxt shows which plugin or module has problem?

**Reply by pi0** [reply]
Yes, we try to use the function name in the warning message but only if it is provided. Otherwise, you need to guess by yourself (first check your local plugins and then toggle used modules one by one to debug)

**Reply by RSS1102** [reply]
> nuxt 可以显示哪个插件或模块有问题吗？
I also look forward to the realization of this function

**Comment by ThibaultVlacich** [comment]
I'm getting this error when running the force upgrade:

```
# npx nuxi@latest upgrade --force
Need to install the following packages:
  nuxi@3.0.0-rc.5
Ok to proceed? (y) y
Nuxt CLI v3.0.0-rc.5                                                                                                                                                                       22:01:22
i Package Manager: yarn 1.22.19                                                                                                                                                            22:01:22

 ERROR  Cannot find module 'nuxt/package.json'                                                                                                                                             22:01:22
Require stack:
- D:\dev\nuxt3

  Require stack:
  - D:\dev\nuxt3
  at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
  at Function.resolve (node:internal/modules/cjs/helpers:108:19)
  at resolveModule (/C:/Users/thiba/AppData/Local/npm-cache/_npx/b95349761371180e/node_modules/nuxi/dist/chunks/cjs.mjs:10:29)
  at getNuxtVersion (/C:/Users/thiba/AppData/Local/npm-cache/_npx/b95349761371180e/node_modules/nuxi/dist/chunks/upgrade.mjs:19:19)
  at Object.invoke (/C:/Users/thiba/AppData/Local/npm-cache/_npx/b95349761371180e/node_modules/nuxi/dist/chunks/upgrade.mjs:41:34)
  at _main (/C:/Users/thiba/AppData/Local/npm-cache/_npx/b95349761371180e/node_modules/nuxi/dist/cli.mjs:46:30)
```

**Comment by ram-you** [comment]
I wonder if there is a plan to implement **extractCSS for Vite** as it exists for Webpack ?

**Reply by ram-you** [reply]
Nobody is interested in this feature?

**Reply by danielroe** [reply]
That is the default behaviour.

**Reply by ram-you** [reply]
As mentioned in docs https://v3.nuxtjs.org/api/configuration/nuxt.config#extractcss
extractCSS is under webpack option in nuxt config. What about vite ?

I would like to be able to set extractcss to false.

**Reply by danielroe** [reply]
Follow https://github.com/nuxt/framework/issues/6755. Neither webpack nor vite support inlining CSS.

**Reply by ram-you** [reply]
@danielroe Thank you very much for these clarifications.

**Comment by Amirali-Yavari** [comment]
hello my dear pooya 
when i  upgrade my nuxt version this error apear 

`Cannot set properties of undefined (setting 'pinia')` 
@pi0

**Reply by mamlzy** [reply]
I got the same error too, how to fix this? @pi0 
https://github.com/nuxt/framework/issues/5955#issue-1306967934

**Reply by ram-you** [reply]
Try upgrade "pinia" and  "@pinia/nuxt", and don't forget  modules: ['@pinia/nuxt'] in nuxt.config.js file.

**Reply by Amirali-Yavari** [reply]
i upgrade pinia but its not work
and has new error
`getActivePinia was called with no active Pinia. Did you forget to install pinia? const pinia = createPinia() app.use(pinia) This will fail in production.`

**Reply by mamlzy** [reply]
> hello my dear pooya when i upgrade my nuxt version this error apear
> 
> `Cannot set properties of undefined (setting 'pinia')` @pi0


I finally fix this after upgrading `@pinia/nuxt` package to the latest version.

**Reply by zavvla** [reply]
hi, there was the same problem, solved it like this:
1. Updated nodejs to min. v14.20.0 (I had a smaller version)
2. Reinstalled the packages https://pinia.vuejs.org/ssr/nuxt.html#installation
3. Restarted the build
![image](https://user-images.githubusercontent.com/11782509/179457824-1f9f96bc-62ba-4fd6-8347-4abc34c21477.png)

**Comment by zavvla** [comment]
Hi, I see that the release date has changed from mid-summer 2022 to summer?

**Reply by danielroe** [reply]
yes. people were misinterpreting it as meteorological midsummer - which has a precise date in June 🤦

**Reply by scherii** [reply]
Oh, so actual mid-summer (around week 32) is still on track? Either way, I really appreciate your work! 🙏

**Comment by pi0** [comment]
Thanks for all feedbacks. Since this is a previous release, I'm locking discussion. Please use a new issue or discussion (or if related to release, latest RC discussion).
